### PR TITLE
[JENKINS-36734] Implement the runSelector Pipeline step

### PR DESCRIPTION
**Pull request by alexsomai**
Related to [[JENKINS-36734]](https://issues.jenkins-ci.org/browse/JENKINS-36734)

Summary of this pull request:
- [x] Implement the `runSelector` Pipeline step
- [x] Write unit tests for the `runSelector` step

CC @ikedam @oleg-nenashev

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -31,13 +31,24 @@
     </developers>
 
     <properties>
-        <jenkins.version>1.580.1</jenkins.version>
+        <jenkins.version>1.590</jenkins.version>
```
Comment (oleg-nenashev) [inline pom.xml]: It would be better to use LTS versions only

File: pom.xml [hunk 2]
```diff
@@ -31,13 +31,24 @@
...
         <java.level>6</java.level>
         <workflow.version>1.4.2</workflow.version>
     </properties>
 
     <dependencies>
         <dependency>
+            <groupId>org.jenkins-ci.plugins.workflow</groupId>
+            <artifactId>workflow-support</artifactId>
+            <version>${workflow.version}</version>
```
Comment (oleg-nenashev) [inline pom.xml:42]: Do you really need it as a mandatory dependency?

Comment (alexsomai) [inline pom.xml:42]: I need this dependency for the `RunWrapper` object. But it is possible to work even if I provide it as an optional dependency.

File: src/test/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStepTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import static java.lang.String.format;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for the {@link RunSelectorStep}.
+ *
+ * @author Alexandru Somai
+ */
+public class RunSelectorStepTest {
```
Comment (alexsomai) [inline src/test/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStepTest.java:28]: These are just some basic tests.
The tests for each selector will be in separate classes, e.g. SpecificRunSelectorTest. But that will be in different PRs.

File: pom.xml [hunk 4]
```diff
@@ -30,13 +30,25 @@
     </developers>
 
     <properties>
-        <jenkins.version>1.580.1</jenkins.version>
+        <jenkins.version>1.596.1</jenkins.version>
```
Comment (ikedam) [inline pom.xml]: JFYI:
I suppose you migrated to Jenkins 1.590 for `Jenkins#getActiveInstance`, which suppresses errors of findbugs for `CheckForNull` annotating`Jenkins#getInstance`.
`Jenkins#getActiveInstance` is deprecated in Jenkins-1.653 and `Jenkins#getInstance` is annotated with `CheckForNull` now. `Jenkins#getInstanceOrNull` is introduced instead.

**Comment by ikedam** [comment]
You might want to set locales in `RunSelectorStepTest` (example: https://github.com/jenkinsci/jenkins/blob/jenkins-2.13/core/src/test/java/hudson/model/RunTest.java#L131)

In my opinion, testing log messages isn't good idea as those tests are easily broken when you updated texts.

**Comment by ikedam** [comment]
Looks good to me.

File: pom.xml (hunk 4 above)
Comment (alexsomai) [inline pom.xml]: Yes, that's why I've upgraded to Jenkins > 1.590. It saves me from one extra null check.
Thanks for the tip, when the time comes to upgrade to Jenkins > 1.653, I'll replace `getActiveInstance` with `getInstance`.

File: README.md [hunk 5]
```diff
@@ -34,3 +34,57 @@ The implementations for the Run Filter are the followings:
...
+```groovy
+def runWrapper = runSelector projectName: 'upstream-project-name', 
+selector: [$class: 'StatusRunSelector', buildStatus: 'Stable'] 
+```
+
+Or, if you'd like to make use of *Permalink*, you can use:
+
+```groovy
+def runWrapper = runSelector projectName: 'upstream-project-name', 
+selector: [$class: 'PermalinkRunSelector', id: 'lastStableBuild']
```
Comment (oleg-nenashev) [inline README.md]: Maybe not the best example since it's similar to the default behavior

File: src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStep.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * The runSelector step selects a specific run from a given project name based on the given selector
+ * and, optionally, the run filter.
+ *
+ * @author Alexandru Somai
+ */
+public class RunSelectorStep extends AbstractStepImpl {
+
+    @CheckForNull
+    private final String projectName;
```
Comment (alexsomai) [inline src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStep.java]: Should this parameter be called `projectName` or `jobName`?
Opinions @ikedam @oleg-nenashev @martinda?

Comment (martinda) [inline src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStep.java]: No opinion here. Jenkins uses `job` and `project` interchangeably.

Comment (ikedam) [inline src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStep.java]: Both is OK, but `jobName` may be preferred especially for pipeline-plugin.
And it can be simply `job` instead of `jobName` like pipeline-build-step-plugin (https://github.com/jenkinsci/pipeline-build-step-plugin).

File: pom.xml (hunk 4 above)
Comment (oleg-nenashev) [inline pom.xml]: I'm not sure that getActiveInstance() is a good justification for the version bump. You can just implement a Util method in your plugin and restrict this Util API

**Comment by martinda** [comment]
Looks good to me.

File: pom.xml (hunk 4 above)
Comment (alexsomai) [inline pom.xml]: You are right, there is not need to update the core version just for this method.

File: src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStep.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+import javax.annotation.CheckForNull;
+
+/**
+ * The runSelector step selects a specific run from a given project name based on the given selector
+ * and, optionally, the run filter.
+ *
+ * @author Alexandru Somai
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorStep.java:19]: Good practice: Add <code>@since TODO</code>

File: src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorExecution.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import org.jenkinsci.plugins.runselector.selectors.StatusRunSelector;
+import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution;
+import org.jenkinsci.plugins.workflow.steps.StepContextParameter;
+import org.jenkinsci.plugins.workflow.support.steps.build.RunWrapper;
+
+/**
+ * The execution of {@link RunSelectorStep}.
+ *
+ * @author Alexandru Somai
+ */
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/runselector/steps/RunSelectorExecution.java:24]: since as well?

**Comment by oleg-nenashev** [comment]
The code looks good to me. 👍

**Comment by oleg-nenashev** [comment]
re-:+1:
